src/storage/src/client/controller/rehydration.rs

```diff
@@ -0,0 +1,275 @@
+                    }
+                } else {
+                    RehydrationTaskState::Pump
+                }
+            }
+            Err(e) => {
+                warn!("storage host produced error, reconnecting: {e}");
+                RehydrationTaskState::Rehydrate
+            }
+        }
+    }
+
+    fn absorb_command(&mut self, command: &StorageCommand<T>) {
+        match command {
+            StorageCommand::IngestSources(ingestions) => {
+                for ingestion in ingestions {
+                    self.ingestions.insert(ingestion.id, ingestion.clone());
+                    // Initialize the uppers we are tracking
+                    self.uppers.insert(
+                        ingestion.id,
+                        (Antichain::from_elem(T::minimum()), {
+                            let mut frontier = timely::progress::frontier::MutableAntichain::new();
+                            frontier.update_iter(Some((T::minimum(), 1)));
+                            frontier
+                        }),
+                    );
+                }
+            }
+            StorageCommand::AllowCompaction(frontiers) => {
+                for (id, frontier) in frontiers {
+                    if frontier.is_empty() {
+                        self.ingestions.remove(id);
+                        self.uppers.remove(id);
+                    }
+                }
+            }
+        }
+    }
+
+    fn absorb_response(&mut self, response: StorageResponse<T>) -> Option<StorageResponse<T>> {
+        match response {
+            StorageResponse::FrontierUppers(mut list) => {
+                for (id, changes) in list.iter_mut() {
+                    if let Some((frontier, frontiers)) = self.uppers.get_mut(id) {
+                        // Apply changes to replica `replica_id`
+                        frontiers.update_iter(changes.drain());
+                        // We can swap `frontier` into `changes, negated, and then use that to repopulate `frontier`.
+                        // Working
+                        changes.extend(frontier.iter().map(|t| (t.clone(), -1)));
+                        frontier.clear();
+                        for (time1, _neg_one) in changes.iter() {
+                            for time2 in frontiers.frontier().iter() {
+                                frontier.insert(time1.join(time2));
+                            }
+                        }
+                        changes.extend(frontier.iter().map(|t| (t.clone(), 1)));
+                        changes.compact();
+                    } else {
+                        // We should have initialized the uppers when we first absorbed
+                        // a command, if storaged has restarted since then.
+                        //
+                        // If the controller has restarted since then, we should have
+                        // initialized them in the initial `step_rehydrate`
+                    }
```

Comment:
`frontiers` is only a single frontier here since we don't have multiple replicas. I recommend calling them something like `reported` and `tracked`! Also I'd advocate panicking if we see an unexpected `FrontierUppers` response, as it is 100% a storaged-side reconciliation error if we do!

```suggestion
                    if let Some((reported, tracked)) = self.uppers.get_mut(id) {
                        // Apply changes to `tracked` frontier.
                        tracked.update_iter(changes.drain());
                        // We can swap `reported` into `changes`, negated, and then use that to repopulate `reported`.
                        changes.extend(reported.iter().map(|t| (t.clone(), -1)));
                        reported.clear();
                        for (time1, _neg_one) in changes.iter() {
                            for time2 in tracked.frontier().iter() {
                                reported.insert(time1.join(time2));
                            }
                        }
                        changes.extend(reported.iter().map(|t| (t.clone(), 1)));
                        changes.compact();
                    } else {
                        // We should have initialized the uppers when we first absorbed
                        // a command, if storaged has restarted since then.
                        //
                        // If the controller has restarted since then, we should have
                        // initialized them in the initial `step_rehydrate`.
                        panic!("RehydratingStorageClient received FrontierUppers response for absent identifier {id}");
                    }
```